Packages/com.unity.inputsystem/InputSystem/Plugins/XR/XRTemplateBuilder.cs

```diff
@@ -0,0 +1,213 @@
+        {
+            int stringLength = templateName.Length;
+            StringBuilder sanitizedTemplateName = new StringBuilder(stringLength);
+            for (int i = 0; i < stringLength; i++)
+            {
+                char letter = templateName[i];
+                if(Char.IsUpper(letter) || Char.IsLower(letter) || Char.IsDigit(letter) || letter == ':')
+                {
+                    sanitizedTemplateName.Append(letter);
+                }
+            }
+            return sanitizedTemplateName.ToString();
+        }
+
+        internal static string OnFindTemplateForDevice(int deviceId, ref InputDeviceDescription description, string matchedTemplate, IInputRuntime runtime)
+        {
+            // If the system found a matching template, there's nothing for us to do.
+            if (!string.IsNullOrEmpty(matchedTemplate))
+            {
+                return null;
+            }
+
+            // If the device isn't a XRInput, we're not interested.
+            if (description.interfaceName != XRUtilities.kXRInterface)
+            {
+                return null;
+            }
+
+            // If the description doesn't come with a XR SDK descriptor, we're not
+            // interested either.
+            if (string.IsNullOrEmpty(description.capabilities))
+            {
+                return null;
+            }
+
+            // Try to parse the XR descriptor.
+            XRDeviceDescriptor deviceDescriptor;
+            try
+            {
+                deviceDescriptor = XRDeviceDescriptor.FromJson(description.capabilities);
+            }
+            catch (Exception)
+            {
+                return null;
+            }
+
+            for (int i = 0; i < availableTemplates.Count; i++)
+            {
+                string templateMatch = availableTemplates[i](deviceDescriptor);
+                if (templateMatch != null)
+                {
+                    return templateMatch;
+                }
+            }
+
+            string templateName = SanitizeTemplateName(string.Format("{0}::{1}::{2}", XRUtilities.kXRInterface, description.manufacturer, description.product));
+            XRTemplateBuilder template = new XRTemplateBuilder { descriptor = deviceDescriptor };
+            InputSystem.RegisterTemplateFactory(() => template.Build(), templateName, null, description);
+
+            return templateName;
+        }
+
+        public InputTemplate Build()
+        {
+            Type deviceType = null;
+            switch(descriptor.deviceRole)
+            {
+                case EDeviceRole.LeftHanded:
+                case EDeviceRole.RightHanded:
+                    {
+                        deviceType = typeof(XRController);
+                    }
+                    break;
+                default:
+                    {
+                        deviceType = typeof(XRHMD);
+                    }
+                    break;

```

Comment:
What you probably want to do instead is to register `XRController` and `XRHMD` as templates and choose them as a base template.